corehq/blobs/s3db.py

```diff
@@ -22,6 +26,43 @@
 DEFAULT_BULK_DELETE_CHUNKSIZE = 1000
 
 
+def is_not_slow_down_error(err: ClientError) -> bool:
+    # SlowDown error message is, "An error occurred (SlowDown) when
+    # calling the GetObject operation (reached max retries: 4): Please
+    # reduce your request rate." From ClientError source we can see that
+    # err.response['Error']['Code'] == 'SlowDown'
+    is_slow_down_error = (
+        'Error' in err.response
+        and 'Code' in err.response['Error']
+        and err.response['Error']['Code'] == 'SlowDown'
+    )
+    return not is_slow_down_error
+
+
+class BackoffDetails(TypedDict):
+    target: str  # reference to the function or method being invoked
+    args: list  # positional arguments to func
+    kwargs: dict  # keyword arguments to func
+    tries: int  # number of invocation tries so far
+    elapsed: float  # elapsed time in seconds so far
+    wait: Optional[float]  # seconds to wait (on_backoff handler only)
+    value: Optional[Any]  # value triggering backoff (on_predicate decorator only)
+
+
+def backoff_handler(details: BackoffDetails):
+    notify_exception(
+        None, "Backing off after receiving SlowDown error from S3",
+        details=details
+    )
+
+
+def giveup_handler(details: BackoffDetails):
+    notify_exception(
+        None, "Giving up after receiving error from S3",
+        details=details
+    )
+
+
```

Comment:
These utilities should be moved below the main class in this module (`S3BlobDB`). If that is not possible maybe put them in a separate module.

Reply:
Yeah, they have to precede their use in the decorator, so a separate module would be better.